### PR TITLE
[602] Consolidate sources of truth for listening mode category visibilty

### LINKED ISSUE #602: Categories List Screen | Last Category Down Button Is Not Disabled
**Description:**
In the categories list screen, last category (8th) down button is not disabled (SS1). But, when we add a custom category as 9th category, last category down button is disabled (SS2). If we remove last (9th) custom category and having totally 8 categories, last category down button is not disabled (SS3). If we add a custom category and remove a preset category, last category down button is disabled (SS4). 

**Steps To Reproduce:**

1. Go to the `Settings`
2. Select `Categories and Phrase`
3. Add a custom category
4. Remove last custom category
5. Add a custom category
6. Remove a preset category

**Expected behavior:**
Last category down button is disabled.

**Actual behavior:**
Last category down button is not disabled.

**Device Information:**
iPad Pro - Test Device - iOS 15.3.1

**Build:**
1.4.0 - 2369 (TestFlight)

**Screenshots:**
**SS1**
SS1_CategoriesListLastDownButtonIsNotDisabled
**SS2**
SS2
**SS3**
SS3
**SS4**
SS4

### PR DESCRIPTION
**Pull request by Clstroud**
Fixes #602 

Restructures and consolidates the various sources of truth for
1) which categories are visible 
2) which aspects of listening mode are enabled
3) how the feature flag for listening mode interacts with those user-facing features

### CODE REVIEW

**Comment by SashankPatel** [comment]
Test Notes 4/27
- Verified on iPhone 8 simulator on iOS 14.
- Verified listen category is hidden when toggle is disabled.
- Verified category hide/show toggle
- Verified bug 603 - the down button is disabled for the last category
Good to go from test

File: Vocable/AppConfig/ListeningModeConfig.swift [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    private func updateCategoryOrdinalsAfterVisibilityChange() {
+        let ctx = NSPersistentContainer.shared.newBackgroundContext()
+        ctx.perform {
+            do {
+                try Category.updateAllOrdinalValues(in: ctx)
+                try ctx.save()
+            } catch {
+                print("Failed to update ordinals: \(error)")
+            }
+        }
```
Comment (jsmorgan42) [inline Vocable/AppConfig/ListeningModeConfig.swift]:
```suggestion
        let context = NSPersistentContainer.shared.newBackgroundContext()
        context.perform {
            do {
                try Category.updateAllOrdinalValues(in: context)
                try context.save()
            } catch {
                print("Failed to update ordinals: \(error)")
            }
        }
```
Just so we're consistent with every other place we declare a context

File: Vocable/AppConfig/ListeningModeConfig.swift [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    @PublishedValue
+    private(set) var isEnabled: Bool = true
+
+    // Whether the hotword feature is allowed to function
+    // This can vary based on the feature flag AND the user preference
+    @PublishedValue
+    private(set) var isHotWordEnabled: Bool = true
+
+    // The user's preference for whether the overall listening mode feature is enabled
+    @PublishedDefault(.isListeningModeEnabled)
```
Comment (jsmorgan42) [inline Vocable/AppConfig/ListeningModeConfig.swift]: Might as well change the user default key name to `.listeningModeEnabledPreference` as well

File: Vocable/AppConfig/ListeningModeConfig.swift [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    var listeningModeEnabledPreference: Bool = ListenModeFeatureConfiguration.deviceSupportsListeningMode()
+
+    // The user's preference for whether the hotword feature is enabled
+    @PublishedDefault(.isHotWordPermitted)
```
Comment (jsmorgan42) [inline Vocable/AppConfig/ListeningModeConfig.swift]: Same thing here, we can change this to `.hotwordEnabledPreference`

File: Vocable/Features/Settings/EditCategories/EditCategoryDetailViewController.swift [hunk 4]
```diff
@@ -271,8 +271,8 @@ final class EditCategoryDetailViewController: VocableCollectionViewController {
         try? Category.updateAllOrdinalValues(in: context)
         saveContext()
```
Comment (jsmorgan42) [inline Vocable/Features/Settings/EditCategories/EditCategoryDetailViewController.swift:272]: This is unrelated to your changes, but should this also be saved within a background context `perform` block?

File: Vocable/Features/Root/CategoriesCarouselViewController.swift [hunk 5]
```diff
@@ -146,18 +147,7 @@ import Combine
...
-        || !AppConfig.isListeningModeSupported
-        || !AppConfig.listeningModeFeatureFlagEnabled
-        || !AppConfig.isListeningModeEnabled
-        || !SpeechRecognitionController.shared.deviceSupportsSpeech
-
-        if shouldRemoveListeningCategory {
-            predicate &= Predicate(\Category.identifier, notEqualTo: Category.Identifier.listeningMode)
-        }
-        request.predicate = predicate
+        request.predicate = Category.visibleCategoriesPredicate(includeUserHidden: false)
```
Comment (jsmorgan42) [inline Vocable/Features/Root/CategoriesCarouselViewController.swift:150]: So much cleaner 💯

**Comment by CananArikan** [comment]
Test verified:
iPad Pro - Test Device - iOS 15.3.1
iPhone 11 - Simulator - iOS 15.4

**Review by SashankPatel — APPROVED** [review]
Ship it

**Pull request merged**